src/sqlfluff/utils/reflow/reindent.py

```diff
@@ -27,3 +36,792 @@ def deduce_line_indent(raw_segment: RawSegment, root_segment: BaseSegment) -> st
+            # It's a consumed indent.
+            current_indent = (
+                cast(TemplateSegment, indent_seg).source_str.split("\n")[-1] or ""
+            )
+        elif not indent_seg.pos_marker or not indent_seg.is_templated:
+            current_indent = indent_seg.raw
+        else:  # pragma: no cover
+            # It's templated. This shouldn't happen. Segments returned by
+            # _get_indent_segment, should be valid indents (i.e. whitespace
+            # or placeholders for consumed whitespace). This is a bug.
+            raise NotImplementedError(
+                "Unexpected templated indent. Report this as a bug on GitHub."
+            )
+    else:
+        current_indent = ""
+
+    # First handle starting indent.
+    assert "\n" not in current_indent, f"Found newline in indent: {indent_seg}"
+    desired_indent_units = indent_line.desired_indent_units(forced_indents)
+    desired_starting_indent = desired_indent_units * single_indent
+    initial_point = cast(ReflowPoint, elements[indent_points[0].idx])
+
+    if current_indent != desired_starting_indent:
+        reflow_logger.debug(
+            "    Correcting indent @ line %s. Existing indent: %r -> %r",
+            elements[indent_points[0].idx + 1].segments[0].pos_marker.working_line_no,
+            current_indent,
+            desired_starting_indent,
+        )
+        # Initial point gets special handling it it has no newlines.
+        if indent_points[0].idx == 0 and not indent_points[0].is_line_break:
+            new_fixes = [
+                LintFix.delete(seg, description="First line should not be indented.")
+                for seg in initial_point.segments
+            ]
+            new_point = ReflowPoint(())
+        # Placeholder indents also get special treatment
+        else:
+            new_fixes, new_point = initial_point.indent_to(
+                desired_starting_indent,
+                **anchor,  # type: ignore
+            )
+        elements[indent_points[0].idx] = new_point
+        fixes += new_fixes
+
+    closing_balance = indent_points[-1].closing_indent_balance
+    # Edge case for first line of a file (where starting indent must be zero).
+    if indent_points[-1].last_line_break_idx is None:
+        starting_balance = 0
+    else:
+        starting_balance = indent_points[0].closing_indent_balance
+
+    # Then check for new lines. Either on the way up...
+    if closing_balance > starting_balance:
+        if indent_points[-1].indent_trough:
+            closing_trough = (
+                indent_points[-1].initial_indent_balance
+                + indent_points[-1].indent_trough
+            )
+        else:
+            closing_trough = (
+                indent_points[-1].initial_indent_balance
+                + indent_points[-1].indent_impulse
+            )
+
+        # On the way up we're looking for whether the ending balance
+        # was an untaken indent on the way up.
```

Comment:
"On the way up" is repeated at the beginning and end of the sentence. Remove one of them?

Reply:
Yeah - I'll rework all the comments in this area.